<a href="https://colab.research.google.com/github/HPashchenko/modelling_data_analyst/blob/main/Data_hw_3_4_apply%2C_groupby%2C_pivot_table.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [3]:
df = pd.read_csv('drive/MyDrive/Data_analysis/supermarket_sales.csv')

In [4]:
df.head(5)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [5]:
df.groupby(by=['Branch'])['Total'].sum()

Branch
A    106200.3705
B    106197.6720
C    110568.7065
Name: Total, dtype: float64

Найбільш прибутковим є філіал супермаркету - С та сумарний прибуток складає 110568б, 7065

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [6]:
df.groupby(by=['Branch', 'City'])['Total'].sum()

Branch  City     
A       Yangon       106200.3705
B       Mandalay     106197.6720
C       Naypyitaw    110568.7065
Name: Total, dtype: float64

В місті Naypyitaw знаходиться філіал С з найбільшим прибутком 110568.7065

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [7]:
invoices_by_category = pd.pivot_table(df, index = 'Branch', columns = 'Product line', values = 'Invoice ID', aggfunc='count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [8]:
invoices_by_category.sort_values(by='Electronic accessories', ascending=False)[:1]

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Згідно проаналізованих даних вище, можемо зробити попередні аналізи, що закономорність продажів залежить від міста та, припустимо, локального асортименту товарів.

Наступними кроками будде перевірено:

1. Чи спостерігається закономірність/відмінність продажів товару від пори року, наприклад в який місяць найбільше якого виду товару було продано. Чи є співпадіння з святами , наприклад.
2. Яка категорія товару найбільш популярна серед міст?
3. Який відсоток оплат становить, наприклад Cash. Чи є динаміка збільшення оплат, використовуючи, наприклад Credit Card

In [9]:
#групуємо дані та визначаємо кількість товарів по містам та департаментам
df.groupby(by=['Branch', 'City'])['Quantity'].sum()

Branch  City     
A       Yangon       1859
B       Mandalay     1820
C       Naypyitaw    1831
Name: Quantity, dtype: int64

In [10]:
# зведена таблиця кількості категоріі товарів по містам
quantity_by_category = pd.pivot_table(df, index = 'City', columns = 'Product line', values = 'Quantity', aggfunc='sum')
quantity_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
City                                                                            
Mandalay                         316                  297                 270   
Naypyitaw                        333                  342                 369   
Yangon                           322                  263                 313   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
City                                                                    
Mandalay                    320                 295                322  
Naypyitaw                   277                 245                265  
Yangon                      257                 371                333

In [11]:
from timeit import default_timer as timer
# Ковертую  колонку Date та виділяю лише місяць
temp = pd.DatetimeIndex(df['Date'])

In [13]:
df['Month'] = temp.month

In [16]:
df['Month'].replace(1, 'January', inplace=True)
df['Month'].replace(2, 'February', inplace=True)
df['Month'].replace(3, 'March', inplace=True)
df['Month'].replace(4, 'April', inplace=True)
df['Month'].replace(5, 'May', inplace=True)
df['Month'].replace(6, 'June', inplace=True)
df['Month'].replace(7, 'July', inplace=True)
df['Month'].replace(8, 'August', inplace=True)
df['Month'].replace(9, 'September', inplace=True)
df['Month'].replace(10, 'October', inplace=True)
df['Month'].replace(11, 'November', inplace=True)
df['Month'].replace(12, 'December', inplace=True)

In [17]:
df['Month'].nunique()

3

In [26]:
#зведена таблиця кількості товарів, проданих потягом трьох місяців, згідно міст
buy_by_month_pivot = pd.pivot_table(df, index = 'Month', columns = 'City', values = 'Quantity', aggfunc='sum', sort=False)
buy_by_month_pivot

City      Yangon  Naypyitaw  Mandalay
Month                                
January      685        680       600
March        681        614       596
February     493        537       624

In [30]:
buy_per_month = df.groupby(by=['Month', 'City', 'Product line'])['Quantity'].sum()
buy_per_month

Month     City       Product line          
February  Mandalay   Electronic accessories    118
                     Fashion accessories       118
                     Food and beverages        107
                     Health and beauty         102
                     Home and lifestyle         83
                     Sports and travel          96
          Naypyitaw  Electronic accessories    106
                     Fashion accessories        98
                     Food and beverages        115
                     Health and beauty         109
                     Home and lifestyle         54
                     Sports and travel          55
          Yangon     Electronic accessories     89
                     Fashion accessories        79
                     Food and beverages        127
                     Health and beauty          55
                     Home and lifestyle         68
                     Sports and travel          75
January   Mandalay   Electronic access

Попередні висновки наступні. Найбільша кількість товарів різноі категорії було здійснено в січні та лютому. Наші припущені були вірні, оскільки січень - лютий місяці - це святкові місяці та, зазвичай, канікули у дітей. Варто перевірити релігійні свята інишх конфесій та регіональні святкуваня даних міст.

In [31]:
df_1 = df.groupby('City')['Payment'].value_counts(normalize=True).reset_index()
df_1

City      Payment  proportion
0   Mandalay      Ewallet    0.340361
1   Mandalay         Cash    0.331325
2   Mandalay  Credit card    0.328313
3  Naypyitaw         Cash    0.378049
4  Naypyitaw      Ewallet    0.323171
5  Naypyitaw  Credit card    0.298780
6     Yangon      Ewallet    0.370588
7     Yangon         Cash    0.323529
8     Yangon  Credit card    0.305882

In [45]:
# зведена таблиця для визначення способів оплат по місяцям
payment_pivot = pd.pivot_table(df, index = 'Month', columns = 'Payment', values = 'Invoice ID',  aggfunc='count')
payment_pivot

Payment   Cash  Credit card  Ewallet
Month                               
February   112           90      101
January    122          113      117
March      110          108      127

Щодо припущення оптат Credit card, можемо спостерігати не закономірну тенденцію використання даного способу оплати. Але привернула увагу іншиа система оплат - це Ewallet. Спостерігаємо цікаву динаміку